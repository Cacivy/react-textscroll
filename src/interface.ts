export enum Mode {
  vertical = 'vertical',
  horizontal = 'horizontal'
}

export interface TextScrollProp {
  className?: string
  mode: string
  text: Array<string>
  speed?: number
}
