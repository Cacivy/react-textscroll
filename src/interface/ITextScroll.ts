export interface TextScrollProp {
    mode: string // "vertical"
    text: Array<string>
    speed?: number

}
export interface TextScrollState {
    duration: number,
    content_width?: number,
    container_width?: number

}