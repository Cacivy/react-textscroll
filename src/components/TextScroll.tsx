import React from 'react'
import styled from 'styled-components'
import ReactDOM from 'react-dom'
import * as Interface from '../interface/ITextScroll'
class TextScroll extends React.Component<Interface.TextScrollProp, Interface.TextScrollState> {
    container: any
    constructor(props: Interface.TextScrollProp) {
        super(props)
        this.state = {
            duration: 10,
            content_width: 500,
            container_width: 500,
        }

    }
    componentDidMount() {
        let dom = ReactDOM.findDOMNode(this.container)
        let parentElement = dom.parentElement
        let children = dom.children
        let containerWidth = parentElement ? parentElement.clientWidth : 500
        let contentWidth = 0
        for (let i = 0; i < children.length; i++) {
            contentWidth += children[i].scrollWidth
            contentWidth += containerWidth
        }
        contentWidth = contentWidth === 0 ? 500 : contentWidth

        let duration = (this.props.speed || 10) * contentWidth / 500000
        this.setState({
            duration: duration,
            content_width: contentWidth,
            container_width: containerWidth
        })

    }
    render() {
        const Container = styled.div`
        position: relative;
        display: block;
        width: auto;
        left:${this.state.container_width}px;
        animation: changebox ${this.state.duration}s linear;
        animation-play-state: running;
        animation-fill-mode: forwards;
        animation-iteration-count: infinite;
        &:hover {
            animation-play-state: paused;
            cursor:default;
        }
        @keyframes changebox {
            0% {
            transform: translate3d(0, 0, 0);
            }
            100% {
            transform: translate3d( -${this.state.content_width}px, 0, 0);
            }
        } 
        `
        const Item = styled.span`
        display:inline-block;
        margin-right: ${this.state.container_width}px
        `
        return (
            <div className={this.props.className}>
                <Container ref={(div: any) => this.container = div}>
                    {this.props.text.map((e: any, i: number) => {
                        return (
                            <Item key={i} > {e}</Item>
                        )
                    })}
                </Container>
            </div>
        )
    }
}
const StyledTextScroll = styled(TextScroll) `
    width:100%;
    height: 100%;
    overflow: hidden;
    word-break: keep-all;
    white-space: nowrap;
    display:flex;
    align-items:center ;
`
export default StyledTextScroll
