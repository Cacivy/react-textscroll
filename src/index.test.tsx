import React from 'react'
import TextScroll from './index'
import { mount, configure } from 'enzyme'
import { TextScrollProp, Mode } from 'interface'
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'
import 'jest-styled-components'
import renderer from 'react-test-renderer'
// import render from 'preact-render-to-json'

configure({ adapter: new Adapter() })

describe('render', () => {
  const data: string[] = [
    '还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：',
    '向上滚动动画',
    '向上滚'
  ]
  it('Vertical should be render', () => {
    
    const props: TextScrollProp = { mode: Mode.vertical, text: data, speed: 5000 }
    const wrapper = mount(<TextScroll {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()

  })
  it(' Horizontal should be render', () => {
    const props: TextScrollProp = { mode: Mode.horizontal, text: data, speed: 5000 }
    const wrapper = mount(<TextScroll {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()

  })
})