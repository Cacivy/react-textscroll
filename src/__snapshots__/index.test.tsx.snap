// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render  Horizontal should be render 1`] = `
.c0 {
  width: 100%;
  height: 40px;
  overflow: hidden;
  border-bottom: 1px solid #eee;
  word-break: keep-all;
  white-space: nowrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
  height: 30px;
  left: 0px;
  -webkit-animation: changebox 5s linear infinite;
  animation: changebox 5s linear infinite;
  -webkit-animation-play-state: running;
  animation-play-state: running;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

.c1:hover {
  -webkit-animation-play-state: paused;
  animation-play-state: paused;
  cursor: default;
}

.c2 {
  display: inline-block;
  line-height: 30px;
  margin-right: 0px;
}

<TextScroll
  mode="horizontal"
  speed={5000}
  text={
    Array [
      "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
      "向上滚动动画",
      "向上滚",
      "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
    ]
  }
>
  <Styled(Horizontal)
    speed={5000}
    text={
      Array [
        "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
        "向上滚动动画",
        "向上滚",
        "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
      ]
    }
  >
    <Horizontal
      className="c0"
      speed={5000}
      text={
        Array [
          "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
          "向上滚动动画",
          "向上滚",
          "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
        ]
      }
    >
      <div
        className="c0"
      >
        <styled.div>
          <div
            className="c1"
          >
            <styled.span
              key="0"
              title="还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码："
            >
              <span
                className="c2"
                title="还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码："
              >
                 
                还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：
              </span>
            </styled.span>
            <styled.span
              key="1"
              title="向上滚动动画"
            >
              <span
                className="c2"
                title="向上滚动动画"
              >
                 
                向上滚动动画
              </span>
            </styled.span>
            <styled.span
              key="2"
              title="向上滚"
            >
              <span
                className="c2"
                title="向上滚"
              >
                 
                向上滚
              </span>
            </styled.span>
            <styled.span
              key="3"
              title="还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码："
            >
              <span
                className="c2"
                title="还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码："
              >
                 
                还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：
              </span>
            </styled.span>
          </div>
        </styled.div>
      </div>
    </Horizontal>
  </Styled(Horizontal)>
</TextScroll>
`;

exports[`render Vertical should be render 1`] = `
.c0 {
  height: 40px;
  border-bottom: 1px solid #eee;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: flex-start;
  align-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  width: 100%;
  height: 30px;
  position: relative;
  top: -0px;
  padding: 0px;
  margin: 0px;
}

.c1 {
  width: 100%;
  overflow: hidden;
  cursor: default;
}

.c3 {
  padding: 0 5px;
  line-height: 30px;
  color: rgba(0,0,0,0.85);
  list-style-type: none;
  cursor: default;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  display: block;
}

<TextScroll
  mode="vertical"
  speed={5000}
  text={
    Array [
      "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
      "向上滚动动画",
      "向上滚",
      "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
    ]
  }
>
  <Styled(Vertical)
    speed={5000}
    text={
      Array [
        "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
        "向上滚动动画",
        "向上滚",
        "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
      ]
    }
  >
    <Vertical
      className="c0"
      speed={5000}
      text={
        Array [
          "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
          "向上滚动动画",
          "向上滚",
          "还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：",
        ]
      }
    >
      <div
        className="c0"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <styled.div>
          <div
            className="c1"
          >
            <styled.ul>
              <ul
                className="c2"
              >
                <styled.li
                  key="0"
                  title="还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码："
                >
                  <li
                    className="c3"
                    title="还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码："
                  >
                    还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：
                  </li>
                </styled.li>
                <styled.li
                  key="1"
                  title="向上滚动动画"
                >
                  <li
                    className="c3"
                    title="向上滚动动画"
                  >
                    向上滚动动画
                  </li>
                </styled.li>
                <styled.li
                  key="2"
                  title="向上滚"
                >
                  <li
                    className="c3"
                    title="向上滚"
                  >
                    向上滚
                  </li>
                </styled.li>
                <styled.li
                  key="3"
                  title="还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码："
                >
                  <li
                    className="c3"
                    title="还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码："
                  >
                    还是上面的例子，我们将transition属性合并，并扩展几个浏览器，如下CSS代码：
                  </li>
                </styled.li>
              </ul>
            </styled.ul>
          </div>
        </styled.div>
      </div>
    </Vertical>
  </Styled(Vertical)>
</TextScroll>
`;
